src/main/java/io/nats/client/impl/MessageQueue.java

```diff
@@ -76,162 +68,104 @@ boolean isDraining() {
 
     void pause() {
         this.running.set(STOPPED);
-        signalAll();
+        this.poisonTheQueue();
     }
 
     void resume() {
         this.running.set(RUNNING);
-        signalAll();
     }
 
     void drain() {
         this.running.set(DRAINING);
-        signalAll();
+        this.poisonTheQueue();
     }
 
     boolean isDrained() {
+        // poison pill is not included in the length count, or the size
         return this.running.get() == DRAINING && this.length() == 0;
     }
 
-    void signalOne() {
-        Thread t = waiters.poll();
-        if (t != null) {
-            LockSupport.unpark(t);
-        }
-    }
-
-    void signalIfNotEmpty() {
-        if (this.length.get() > 0) {
-            signalOne();
-        }
-    }
-
-    void signalAll() {
-        Thread t = waiters.poll();
-        while(t != null) {
-            LockSupport.unpark(t);
-            t = waiters.poll();
-        }
-    }
-
     void push(NatsMessage msg) {
-
         // If we aren't running, then we need to obey the filter lock
         // to avoid ordering problems
         if(!this.isRunning()) {
             this.filterLock.lock();
-            this.queue.add(msg);
+            this.put(msg);
             this.filterLock.unlock();
             this.sizeInBytes.getAndAdd(msg.getSizeInBytes());
             this.length.incrementAndGet();
-            signalOne();
             return;
         }
 
-        this.queue.add(msg);
+        this.put(msg);
         this.sizeInBytes.getAndAdd(msg.getSizeInBytes());
         this.length.incrementAndGet();
-        signalOne();
     }
 
-    NatsMessage waitForTimeout(Duration timeout) throws InterruptedException {
-        long timeoutNanos = (timeout != null) ? timeout.toNanos() : -1;
-        NatsMessage retVal = null;
-
-        if (timeoutNanos >= 0) {
-            Thread t = Thread.currentThread();
-            long start = System.nanoTime();
+    void poisonTheQueue() {
+        try {
+            this.queue.add(this.poisonPill);
+        } catch (IllegalStateException ie) { // queue was full, so we don't really need poison pill
+            // ok to ignore this
+        }
+    }
 
-            // Semi-spin for at most MAX_SPIN_TIME
-            if (timeoutNanos > MAX_SPIN_TIME) {
-                int count = 0;
-                while (this.isRunning() && (retVal = this.queue.poll()) == null && count < this.maxSpins) {
+    void put(NatsMessage msg) {
+        try {
+            this.queue.put(msg);
+        } catch (InterruptedException ie) {
+            // ok to ignore this

```

Comment:
Why "ok", will this not put the msg on the queue? Is this assumed to always be terminating the process?

Reply:
added comment